src/compiler/transformers/classFields.ts

```diff
@@ -411,8 +449,89 @@ namespace ts {
             return statement;
         }
 
-        function visitComputedPropertyName(node: ComputedPropertyName) {
-            let expression = visitNode(node.expression, visitor, isExpression);
+        function visitVariableDeclaration(node: VariableDeclaration) {
+            // 14.3.1.2 RS: Evaluation
+            //   LexicalBinding : BindingIdentifier Initializer
+            //     ...
+            //     3. If IsAnonymousFunctionDefinition(|Initializer|) is *true*, then
+            //        a. Let _value_ be ? NamedEvaluation of |Initializer| with argument _bindingId_.
+            //     ...
+            //
+            // 14.3.2.1 RS: Evaluation
+            //   VariableDeclaration : BindingIdentifier Initializer
+            //     ...
+            //     3. If IsAnonymousFunctionDefinition(|Initializer|) is *true*, then
+            //        a. Let _value_ be ? NamedEvaluation of |Initializer| with argument _bindingId_.
+            //     ...
+
+            if (isIdentifier(node.name) && node.initializer &&
+                isAnonymousClassDeclarationNeedingAssignedName(node.initializer)) {
+                const original = getOriginalNode(node.initializer, isClassLike);
+                const referencedName = original && !original.name && hasSyntacticModifier(original, ModifierFlags.Default) ?
+                    factory.createStringLiteral("default") :
+                    factory.createStringLiteralFromNode(node.name);
+                const name = visitNode(node.name, visitor, isBindingName);

```

Comment:
why is the test provided to `visitNode` `isBindingName`?

Reply:
That is the type allowed by `updateVariableDeclaration`. `name` probably won't end up transformed into a binding pattern by `visitor`, but it is permitted nonetheless.